File: BDA_frontend/src/services/PuntoInteresService.js
import api from './api';

const RESOURCE_URL = '/api/puntos-interes';

class PuntoInteresService {
  // CRUD básico
  getAll() {
    return api.get(RESOURCE_URL);
  }

  getAllPaginated(page = 0, size = 10) {
    return api.get(`${RESOURCE_URL}/paginated`, { params: { page, size } });
  }

  getById(id) {
    return api.get(`${RESOURCE_URL}/${id}`);
  }

  create(punto) {
    return api.post(RESOURCE_URL, punto);
  }

  update(id, punto) {
    return api.put(`${RESOURCE_URL}/${id}`, punto);
  }

  delete(id) {
    return api.delete(`${RESOURCE_URL}/${id}`);
  }

  // Queries simples
  getByTipo(tipo) {
    return api.get(`${RESOURCE_URL}/tipo/${tipo}`);
  }

  getByZona(idZona) {
    return api.get(`${RESOURCE_URL}/zona/${idZona}`);
  }

  getByNombre(nombre) {
    return api.get(`${RESOURCE_URL}/nombre`, { params: { nombre } });
  }

  getNearby(latitud, longitud, radioKm = 5.0) {
    return api.get(`${RESOURCE_URL}/cercanos`, { params: { latitud, longitud, radioKm } });
  }

  // Alias en español
  obtenerPaginado(page, size) {
    return this.getAllPaginated(page, size);
  }

  crear(punto) {
    return this.create(punto);
  }

  actualizar(id, punto) {
    return this.update(id, punto);
  }

  eliminar(id) {
    return this.delete(id);
  }
}

export default new PuntoInteresService();

File: BDA_backend/src/main/java/com/example/bda/Services/PuntoInteresService.java
package com.example.bda.Services;

import com.example.bda.Model.puntos_interes;
import com.example.bda.Repository.PuntoInteresRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Service
public class PuntoInteresService {

    @Autowired
    private PuntoInteresRepository repository;

    // CREATE
    public puntos_interes create(puntos_interes punto) {
        int id = repository.create(punto);
        punto.setId_punto(id);
        return punto;
    }

    // READ - Obtener todos
    public List<puntos_interes> getAll() {
        return repository.findAll();
    }

    // READ - Obtener con paginación
    public Map<String, Object> getAllPaginated(int page, int size) {
        List<puntos_interes> puntos = repository.findAllPaginated(page, size);
        int total = repository.count();
        
        Map<String, Object> response = new HashMap<>();
        response.put("puntos", puntos);
        response.put("currentPage", page);
        response.put("totalItems", total);
        response.put("totalPages", (int) Math.ceil((double) total / size));
        return response;
    }

    // READ - Obtener por ID
    public puntos_interes getById(int id) {
        return repository.findById(id);
    }

    // UPDATE
    public puntos_interes update(int id, puntos_interes punto) {
        int rowsAffected = repository.update(id, punto);
        if (rowsAffected > 0) {
            punto.setId_punto(id);
            return punto;
        }
        return null;
    }

    // DELETE
    public boolean delete(int id) {
        return repository.delete(id) > 0;
    }

    // QUERIES SIMPLES
    public List<puntos_interes> getByTipo(String tipo) {
        return repository.findByTipo(tipo);
    }

    public List<puntos_interes> getByZona(int idZona) {
        return repository.findByZona(idZona);
    }

    public List<puntos_interes> getByNombre(String nombre) {
        return repository.findByNombre(nombre);
    }

    public List<puntos_interes> getNearby(double latitud, double longitud, double radioKm) {
        return repository.findNearby(latitud, longitud, radioKm);
    }
}
